File: src/main/java/com/codingforcookies/betterrecords/item/ItemWire.java
package com.codingforcookies.betterrecords.item;

import com.codingforcookies.betterrecords.api.connection.RecordConnection;
import com.codingforcookies.betterrecords.api.wire.IRecordWire;
import com.codingforcookies.betterrecords.api.wire.IRecordWireHome;
import com.codingforcookies.betterrecords.api.wire.IRecordWireManipulator;
import com.codingforcookies.betterrecords.common.core.helper.ConnectionHelper;
import com.codingforcookies.betterrecords.common.packets.PacketHandler;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.EnumFacing;
import net.minecraft.world.World;

public class ItemWire extends BetterItem implements IRecordWireManipulator {

    public static RecordConnection connection;

    public ItemWire(String name) {
        super(name);
    }

    @Override
    public EnumActionResult onItemUse(ItemStack stack, EntityPlayer player, World world, BlockPos pos, EnumHand hand, EnumFacing side, float hitX, float hitY, float hitZ) {
        if(!world.isRemote)
            return EnumActionResult.PASS;

        TileEntity te = world.getTileEntity(pos);
        if(te == null || !(te instanceof IRecordWire))
            return EnumActionResult.PASS;

        if(connection == null) {
            connection = new RecordConnection(pos.getX(), pos.getY(), pos.getZ(), te instanceof IRecordWireHome);
        }else{
            float x1 = -(float)(pos.getX() - (connection.fromHome ? connection.x1 : connection.x2));
            float y1 = -(float)(pos.getY() - (connection.fromHome ? connection.y1 : connection.y2));
            float z1 = -(float)(pos.getZ() - (connection.fromHome ? connection.z1 : connection.z2));

            if((int)Math.sqrt(Math.pow(x1, 2) + Math.pow(y1, 2) + Math.pow(z1, 2)) > 7 || connection.sameInitial(pos.getX(), pos.getY(), pos.getZ())) {
                connection = null;
                return EnumActionResult.PASS;
            }

            if(!connection.fromHome)
                connection.setConnection1(pos.getX(), pos.getY(), pos.getZ());
            else
                connection.setConnection2(pos.getX(), pos.getY(), pos.getZ());

            TileEntity te1 = world.getTileEntity(new BlockPos(connection.x1, connection.y1, connection.z1));
            TileEntity te2 = world.getTileEntity(new BlockPos(connection.x2, connection.y2, connection.z2));

            if(te2 instanceof IRecordWire) {
                if(!(te1 instanceof IRecordWireHome && te2 instanceof IRecordWireHome)) {
                    ConnectionHelper.addConnection(te.getWorld(), (IRecordWire)te1, connection, world.getBlockState(te.getPos()));
                    ConnectionHelper.addConnection(te.getWorld(), (IRecordWire)te2, connection, world.getBlockState(te.getPos()));
                    PacketHandler.sendWireConnectionFromClient(connection);
                    stack.stackSize--;
                }
            }

            connection = null;
        }
        return EnumActionResult.PASS;
    }
}

File: src/main/kotlin/com/codingforcookies/betterrecords/item/ItemWire.kt
package com.codingforcookies.betterrecords.item

import com.codingforcookies.betterrecords.api.connection.RecordConnection
import com.codingforcookies.betterrecords.api.wire.IRecordWire
import com.codingforcookies.betterrecords.api.wire.IRecordWireHome
import com.codingforcookies.betterrecords.api.wire.IRecordWireManipulator
import com.codingforcookies.betterrecords.common.core.helper.ConnectionHelper
import com.codingforcookies.betterrecords.common.packets.PacketHandler
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

class ItemWire(name: String) : ModItem(name), IRecordWireManipulator {

    companion object {
        var connection: RecordConnection? = null
    }

    override fun onItemUse(stack: ItemStack, player: EntityPlayer, world: World, pos: BlockPos, hand: EnumHand, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): EnumActionResult {
        if (!world.isRemote) {
            return EnumActionResult.PASS
        }

        (world.getTileEntity(pos) as? IRecordWire)?.let { te ->
            connection?.let {

                val x1 = -(pos.x - if (it.fromHome) it.x1 else it.x2).toFloat()
                val y1 = -(pos.y - if (it.fromHome) it.y1 else it.y2).toFloat()
                val z1 = -(pos.z - if (it.fromHome) it.z1 else it.z2).toFloat()

                if (Math.sqrt(Math.pow(x1.toDouble(), 2.toDouble()) + Math.pow(y1.toDouble(), 2.toDouble()) + Math.pow(z1.toDouble(), 2.toDouble())) > 7 || it.sameInitial(pos.x, pos.y, pos.z)) {
                    connection = null
                    return EnumActionResult.PASS
                }
                
                if (!it.fromHome) {
                    it.setConnection1(pos.x, pos.y, pos.z)
                } else {
                    it.setConnection2(pos.x, pos.y, pos.z)
                }

                val te1 = world.getTileEntity(BlockPos(it.x1, it.y1, it.z1))
                val te2 = world.getTileEntity(BlockPos(it.x2, it.y2, it.z2))

                if (te2 is IRecordWire) {
                    if (!(te1 is IRecordWireHome && te2 is IRecordWireHome)) {
                        ConnectionHelper.addConnection((te as TileEntity).world, te1 as IRecordWire, connection, world.getBlockState(te.getPos()))
                        ConnectionHelper.addConnection((te as TileEntity).world, te2 as IRecordWire, connection, world.getBlockState(te.getPos()))
                        PacketHandler.sendWireConnectionFromClient(connection)
                        stack.stackSize--
                    }
                }

                return EnumActionResult.PASS
            }

            connection = RecordConnection(pos.x, pos.y, pos.z, te is IRecordWireHome)
        }

        return EnumActionResult.PASS
    }
}
